File: 08/08.jl
timestamp_nano() = Int(round(time() * 1e9))

function part1(grid)::Int
    # no double counting in the corners
    n = size(grid, 1) * 2 + size(grid, 2) * 2 - 4
    for x in 2:(size(grid, 1)-1)
        for y in 2:(size(grid, 2)-1)
            tree = grid[x, y]
            n += all(tree .> grid[x+1:end, y]) || all(tree .> grid[1:x-1, y]) || all(tree .> grid[x, y+1:end]) || all(tree .> grid[x, 1:y-1])
        end
    end
    return n
end

function part2(grid)::Int
    scenic_score = 0
    h, w = size(grid, 1) - 1, size(grid, 2) - 1

    for i in 2:h
        for j in 2:w
            tree = grid[i, j]

            score_up = 1
            x = i
            while x > 2 && tree > grid[x-=1, j]
                score_up += 1
            end

            score_left = 1
            y = j
            while y > 2 && tree > grid[i, y-=1]
                score_left += 1
            end

            score_down = 1
            x = i
            while x < w - 1 && tree > grid[x+=1, j]
                score_down += 1
            end

            score_right = 1
            y = j
            while y < h - 1 && tree > grid[i, y+=1]
                score_right += 1
            end

            scenic_score = max(score_up * score_left * score_down * score_right, scenic_score)
        end
    end
    return scenic_score
end

get_data(path::String) =
    open(path) do file
        content = read(file, String)
        lines = split(content, "\n")

        reduce(hcat, parse.(Int, split(x, "")) for x in lines)'
    end

start = timestamp_nano()

grid = get_data("08/input.txt")

println("part1: ", part1(grid))
println("part2: ", part2(grid))

end_ = timestamp_nano()
println("time: ", round((end_ - start) / 1e6, digits=3), "ms")

File: 08/08.py
from timing_util import Timing


def get_data(content: str) -> list[list[int]]:
    return [list(map(int, line)) for line in content.splitlines()]


def part1(grid: list[list[int]]) -> int:
    # no double counting on corners
    circumference = len(grid) * 2 + len(grid[0]) * 2 - 4

    n_visible = circumference
    for i, row in enumerate(grid[1:-1], 1):
        for j, tree in enumerate(row[1:-1], 1):
            n_visible += all(tree > r[j] for r in grid[:i]) or \
                all(tree > r[j] for r in grid[i+1:]) or \
                all(tree > t for t in row[:j]) or \
                all(tree > t for t in row[j+1:])

    return n_visible


def part2(grid: list[list[int]]) -> int:
    scenic_score = 0
    for i, row in enumerate(grid[1:-1], 1):
        for j, tree in enumerate(row[1:-1], 1):
            score_up = 1
            for r in grid[1:i][::-1]:
                if tree <= r[j]:
                    break
                score_up += 1

            score_left = 1
            for t in row[1:j][::-1]:
                if tree <= t:
                    break
                score_left += 1

            score_down = 1
            for r in grid[i + 1:-1]:
                if tree <= r[j]:
                    break
                score_down += 1

            score_right = 1
            for t in row[j + 1:-1]:
                if tree <= t:
                    break
                score_right += 1

            scenic_score = max(scenic_score, score_up * score_left * score_down * score_right)
    return scenic_score


if __name__ == '__main__':
    with Timing():
        with open('08/input.txt') as in_file:
            data = get_data(in_file.read())

        print(f'part1: {part1(data)}')
        print(f'part2: {part2(data)}')
